addon/record-data.js

```diff
@@ -149,7 +150,7 @@ export default class FragmentRecordData extends RecordData {
     } else if (isFragment(value)) {
       // A fragment instance was given, so just replace the existing value
       fragment = setFragmentOwner(value, record, key);
-    } else if (!fragment) {
+    } else if (!fragment || (getActualFragmentType(declaredModelName, options, value, record) !== fragment.constructor.modelName)) {
```

Comment:
I think we need a comment here, or maybe put this into a descriptive variable so we know what this is doing in the future.